strawberryfields/engine.py

```diff
@@ -218,45 +227,51 @@ def _broadcast_nones(val, shots):
         if not isinstance(program, Sequence):
             program = [program]
 
-        try:
-            prev = self.run_progs[-1] if self.run_progs else None  # previous program segment
-            for p in program:
-                if prev is None:
-                    # initialize the backend
-                    self._init_backend(p.init_num_subsystems)
-                else:
-                    # there was a previous program segment
-                    if not p.can_follow(prev):
-                        raise RuntimeError("Register mismatch: program {}, '{}'.".format(len(self.run_progs), p.name))
-
-                    # Copy the latest measured values in the RegRefs of p.
-                    # We cannot copy from prev directly because it could be used in more than one engine.
-                    for k, v in enumerate(self.samples):
-                        p.reg_refs[k].val = v
-
-                # if the program hasn't been compiled for this backend, do it now
-                if p.backend != self.backend_name:
-                    p = p.compile(self.backend_name, **compile_options) # TODO: shots might be relevant for compilation?
-                p.lock()
-
-                kwargs["shots"] = shots
-                # Note: by putting ``shots`` into keyword arguments, it allows for the
-                # signatures of methods in Operations to remain cleaner, since only
-                # Measurements need to know about shots
-
-                self._run_program(p, **kwargs)
-                self.run_progs.append(p)
-                # store the latest measurement results
-                shots = kwargs.get("shots", 1)
-                self.samples = [_broadcast_nones(p.reg_refs[k].val, shots) for k in sorted(p.reg_refs)]
-                prev = p
-        except Exception as e:
-            raise e
-        else:
-            # program execution was successful
-            pass
-
-        return Result(self.samples.copy())
+        kwargs["shots"] = shots
+        # NOTE: by putting ``shots`` into keyword arguments, it allows for the
+        # signatures of methods in Operations to remain cleaner, since only
+        # Measurements need to know about shots
+
+        if self.backend_name in getattr(self, "HARDWARE_BACKENDS", []):
+            p = program[0]
+            p = p.compile(self.backend_name)  # TODO: does compile need to know about shots?
+            p.lock()
+            self.run_progs.append(p)
+            samples = self._run_program(p, **kwargs)
+            return Result(samples)
```

Comment:
I guess the idea here is that hardware backends can only run a single program before they have to be reset? If so, maybe `DeviceSpecs.interactive` should be checked instead.

Also, it would be possible to combine the Programs on the client side before execution.

Reply:
Using `DeviceSpecs.interactive` during compilation sounds like the right approach.

Reply:
Ok let me look into this.

Reply:
@smite the only way to combine programs with the current hardware support, as I understand it, would be to concatenate all the programs and ensure that there is one measurement at the end. This is out of scope for this PR but it is something I discussed with @co9olguy briefly.

Reply:
@smite @josh146 using `DeviceSpecs.interactive` here breaks some existing tests (e.g. `test_print_applied`) because the backend fixture in those tests uses the `base` backend which is set to interactive. I assume this is intentional, but tests that run two programs will have to be updated. E.g.
```
        # ... line 162 in test_engine.py
        # reapply history
        eng.reset()
        eng.run([p1, p2])
        assert inspect() == expected2
```

Reply:
Checking the interactivity of the backend during compilation probably does not make much sense, since the Programs are compiled individually and don't know about each other...

Reply:
That check fails if it's in the `_run` method, before any compiling happens, but because those tests are passing multiple programs but the base backend is set to `interactive`. Likely if these tests expect the backend to not be interactive, they should initialize one with that setting.